### PR TITLE
Algorithmic improvements to AL mortar mechanical contact

### LINKED ISSUE #25404: Improve mortar AL contact
## Reason
A few things in AL contact require enhancements:

1. The adaptivity of the coefficients depend on the AL iteration (to be avoided)
2. Lack of options for adapting normal and frictional penalty values
3. The LM values start from zero every time step
4. The penalty values can reach large values after AL augments, thereby compromising the solution of the system

## Design
Use existing objects and expose the parameters in the contact action

## Impact
One more step toward industrial-level use of mortar mechanical contact.

### PR DESCRIPTION
**Pull request by recuero**
Closes #25404

### CODE REVIEW

**Comment by recuero** [comment]
Address https://github.com/idaholab/moose/pull/24349#pullrequestreview-1592516437

File: modules/contact/test/tests/3d-mortar-contact/tests [hunk 1]
```diff
@@ -183,4 +183,18 @@
     rel_err = 1.0e-5
     abs_zero = 1.0e-5
   []
+  [frictional-mortar-3d-al]
+    type = 'Exodiff'
+    input = 'frictional-mortar-3d-al.i'
+    exodiff = 'frictional-mortar-3d-al_out.e'
+    requirement = 'The system shall solve a 3D frictional problem with mortar constraint using a '
```
Comment (dschwen) [inline modules/contact/test/tests/3d-mortar-contact/tests:190]:
```suggestion
    requirement = 'The system shall be able to solve a 3D frictional problem with mortar constraint using a '
```

**Comment by recuero** [comment]
Need to add coverage.

**Review by dschwen — COMMENTED** [review]
Looks fine to me, just some minor comments.

File: modules/contact/include/userobjects/PenaltyFrictionUserObject.h [hunk 2]
```diff
@@ -63,6 +63,9 @@ class PenaltyFrictionUserObject : virtual public PenaltyWeightedGapUserObject,
   /// The friction coefficient
   const Real _friction_coefficient;
 
+  /// Map from degree of freedom to current and old step slip
+  std::unordered_map<const DofObject *, std::pair<TwoVector, TwoVector>> _dof_to_step_slip;
```
[the remark this answers is not in the document]
Comment (recuero) [reply modules/contact/include/userobjects/PenaltyFrictionUserObject.h:67]: Agreed. We may see other bottlenecks in the profiler report too.

**Comment by dschwen** [comment]
Merge commit?

**Comment by recuero** [comment]
Topher's residual tags PR created a conflict.

**Comment by recuero** [comment]
I'll rebase locally. This on-PR merge commit is a mess.

**Comment by recuero** [comment]
@dschwen this should be ready for another review or approval.

**Review by dschwen — APPROVED** [review]

**Pull request merged**